### PR TITLE
ZEPPELIN-2645. Adding way to register RemoteInterpreterServer's port into InterpreterProcess

### PR DESCRIPTION
**Pull request by zjffdu**
### What is this PR for?

Rebase PR #2418 ,  still use thrift as the communication protocol between zeppelin server and interpreter process.  We can change it io netty in future when we implement 2 way communication channel between zeppelin server and interpreter process. 

### What type of PR is it?
[Bug Fix | Improvement | Feature | Documentation | Hot Fix | Refactoring]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-2645

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? NO
* Does this needs documentation? No

### CODE REVIEW

**Comment by zjffdu** [comment]
@Leemoonsoo @jongyoul Please help review, thanks

File: zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java [hunk 1]
```diff
@@ -357,16 +357,16 @@ public void testSchedule() throws InterruptedException, IOException {
     config.put("cron", "* * * * * ?");
     note.setConfig(config);
     notebook.refreshCron(note.getId());
-    Thread.sleep(1 * 1000);
+    Thread.sleep(2 * 1000);
```
Comment (zjffdu) [inline zeppelin-zengine/src/test/java/org/apache/zeppelin/notebook/NotebookTest.java]: This is for flaky test

**Comment by jongyoul** [comment]
LGTM. I agree with using Thrift to implement this feature as we don't have to add new libs more.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterManagedProcess.java [hunk 2]
```diff
@@ -96,18 +116,66 @@ public int getPort() {
   @Override
   public void start(String userName, Boolean isUserImpersonate) {
     // start server process
+    final String callbackHost;
+    final int callbackPort;
     try {
       port = RemoteInterpreterUtils.findRandomAvailablePortOnAllLocalInterfaces();
       logger.info("Choose port {} for RemoteInterpreterProcess", port);
+      callbackHost = RemoteInterpreterUtils.findAvailableHostname();
+      callbackPort = RemoteInterpreterUtils.findRandomAvailablePortOnAllLocalInterfaces();
```
Comment (Leemoonsoo) [inline zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterManagedProcess.java]: I think it's not very uncommon that system administrator setup firewall with ports whitelisted.
Once Interpreter is run in remote machines, this randomly assigned port may make firewall setup difficult.

Any thoughts on this?

**Comment by Leemoonsoo** [comment]
Left a question. except for that, LGTM.

File: zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterManagedProcess.java (hunk 2 above)
Comment (zjffdu) [reply zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/remote/RemoteInterpreterManagedProcess.java]: Make sense, will update the PR.

**Comment by zjffdu** [comment]
@Leemoonsoo @jongyoul I implement the portRange in thrift callback server. But didn't do that for the port of `RemoteInterpreterServer`. I will left it when implementing cluster mode.

**Comment by zjffdu** [comment]
Will commit if no more comments.

**Pull request closed without merging**